### ISSUE TITLE
Gateway resources fail due to long time to create in AWS

### ISSUE THREAD

**Issue by jaferrando** [AviatrixSystems/terraform-provider-aviatrix#15]
Creating a gateway takes long as it waits for the gw instance to be booted in AWS. Terraform fails and does not save the resource in the state. A new terraform plan informs that it will in effect try to create the gateway again which in effect would create a second gateway.
Trying a terraform import does not work, as the provider does not support import.
Could the gw creation be asyncronous?

**Comment by jaferrando** [comment]
Gateway resource does not support configurable Terraform timeouts either

**Comment by jaferrando** [comment]
Sample error result creating a gateway:

aviatrix_gateway.sec-vpn-aza-gw-1: Still creating... (10s elapsed)
aviatrix_gateway.sec-vpn-aza-gw-1: Still creating... (20s elapsed)
aviatrix_gateway.sec-vpn-aza-gw-1: Still creating... (30s elapsed)
aviatrix_gateway.sec-vpn-aza-gw-1: Still creating... (40s elapsed)
aviatrix_gateway.sec-vpn-aza-gw-1: Still creating... (50s elapsed)

Error: Error applying plan:

1 error(s) occurred:

* aviatrix_gateway.sec-vpn-aza-gw-1: 1 error(s) occurred:

* aviatrix_gateway.sec-vpn-aza-gw-1: Failed to create Aviatrix Gateway: EOF

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.

**Comment by AviatrixSystems** [comment]
Thanks for your feedback. Multiple gateways cannot be created in parallel. We are working on fixing the issues.

**Comment by edumanig** [comment]
Please do git pull or git clone and give another try.  There is another way to throttle terraform, add this option -parallelism=<n> when doing terraform apply.

**Comment by jalvarezferr** [comment]
After pulling the latest to create a new Gateway, the issue persists.
-parallelism does not help change the result

**Comment by arvindsree** [comment]
Gateway's can be created in parallel(However due to locks on the controller it will take 2x time)
Terraform must not crash because it takes longer time. If the API itself failed for some reason, please enable debugging with TF_LOG=TRACE terraform apply and submit the output with error to Aviatrix Support

**Comment by jalvarezferr** [comment]
I'm not trying to create more than one gateway in parallel. A single gateway resource creation fails every time after exactly 60 seconds:

´´´
2019/01/21 10:41:22 [TRACE] dag/walk: vertex "meta.count-boundary (count boundary fixup)", waiting for: "aviatrix_gateway.sec-vpn-aza-eu-west-1"
2019/01/21 10:41:22 [TRACE] dag/walk: vertex "provider.aviatrix (close)", waiting for: "aviatrix_gateway.sec-vpn-aza-eu-west-1"
2019-01-21T10:41:22.575+0100 [DEBUG] plugin.terraform-provider-aviatrix.exe: 2019/01/21 10:41:22 [INFO] failed to create Aviatrix gateway: &goaviatrix.Gateway{AccountName:"DXCIBER_SEC", Action:"connect_container", AdditionalCidrs:"", AuthMethod:"", AllocateNewEip:"", BkupGatewayZone:"", BkupPrivateIP:"", CID:"qdjZAbvbFe7IBLltKawe", CIDR:"", ClientCertAuth:"", ClientCertSharing:"", CloudType:1, CloudnBkupGatewayInstID:"", CloudnGatewayInstID:"", DirectInternet:"", DockerConsulIP:"", DockerNtwkCidr:"", DockerNtwkName:"", DuoAPIHostname:"", DuoIntegrationKey:"", DuoPushMode:"", DuoSecretKey:"", Eip:"", ElbDNSName:"", ElbName:"", ElbState:"", EnableClientCertSharing:"", EnableElb:"yes", EnableLdap:"", DnsServer:"", PublicDnsServer:"", EnableNat:"", SingleAZ:"", EnablePbr:"", Expiration:"", GatewayZone:"", GwName:"sec-vpn-aza-eu-west-1", GwSecurityGroupID:"", GwSize:"", GwSubnetID:"", HASubnet:"", PeeringHASubnet:"", NewZone:"", InstState:"", IntraVMRoute:"", IsHagw:"", LdapAdditionalReq:"", LdapBaseDn:"", LdapBindDn:"", LdapCaCert:"", LdapClientCert:"", LdapPassword:"", LdapServer:"", LdapUseSsl:"", LdapUserAttr:"", LicenseID:"", MaxConn:"", MaxConnections:"", Nameservers:"", OktaToken:"", OktaURL:"", OktaUsernameSuffix:"", OtpMode:"", PbrDefaultGateway:"", PbrEnabled:"", PbrLogging:"", PbrSubnet:"", PrivateIP:"", PublicIP:"", SamlEnabled:"", SandboxIP:"", SaveTemplate:"", SearchDomains:"", SplitTunnel:"yes", TunnelName:"", TunnelType:"", VendorName:"", VpcID:"vpc-eb7de88d", VpcNet:"172.31.16.0/20", VpcRegion:"eu-west-1", VpcSize:"t2.small", VpcSplunkIPPort:"", VpcState:"", VpcType:"", VpnCidr:"192.168.44.0/24", VpnStatus:"yes", Zone:""}
2019/01/21 10:41:22 [TRACE] root: eval: *terraform.EvalWriteState
2019/01/21 10:41:22 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2019/01/21 10:41:22 [TRACE] root: eval: *terraform.EvalIf
2019/01/21 10:41:22 [TRACE] root: eval: *terraform.EvalWriteState
2019/01/21 10:41:22 [TRACE] root: eval: *terraform.EvalWriteDiff
2019/01/21 10:41:22 [TRACE] root: eval: *terraform.EvalApplyPost
2019/01/21 10:41:22 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* aviatrix_gateway.sec-vpn-aza-eu-west-1: failed to create Aviatrix gateway: EOF
2019/01/21 10:41:22 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* aviatrix_gateway.sec-vpn-aza-eu-west-1: failed to create Aviatrix gateway: EOF
2019/01/21 10:41:22 [TRACE] [walkApply] Exiting eval tree: aviatrix_gateway.sec-vpn-aza-eu-west-1
´´´

As the go-aviatrix.Client functions do not configure timeout on the HTTPClient, this seems either at timeout at the server side or somewhere along my connection. As I have a direct connection without any proxy or other intermediary layer 4+ devices, my first guess is that it is the controller closing the connection.

I have made direct calls to the API with curl and they all end without a response after 60 seconds. Is this a Controller's timeout? Is this only happening to me and should I open a support ticket?

**Comment by jalvarezferr** [comment]
After going though this timeout problem with Aviatrix support and doing my own tests, I've nailed down the reason, begin this that we assigned a DNS name to the Controller instead of using its public IP address. Apparantly the timout configuration in the Controller's Apache is set in a virtual host definition associated to the IP address itself, while the default remains with defaut value wich for this Apache version (seemingly 2.4) is 60 seconds (previous versions used a 300 seconds default).
So, being able to customize the Aviatrix timeout would not help solve this issue, as it turns to happen at the Aviatrix controller and not at Terraform.
We can close this issue.
